input-events/input-events-typing.html

```diff
@@ -1,82 +1,311 @@
-        'plain-keypress-a',
-        'plain-beforeinput-a',
-        'plain-input-a',
-        'plain-keyup-a',
-        // Pressing Shift-'b'.
-        'plain-keydown-B',
-        'plain-keypress-B',
-        'plain-beforeinput-B',
-        'plain-input-B',
-        'plain-keyup-B',
-        // Pressing 'c'.
-        'rich-keydown-c',
-        'rich-keypress-c',
-        'rich-beforeinput-c',
-        'rich-input-c',
-        'rich-keyup-c',
-        // Pressing Shift-'d'.
-        'rich-keydown-D',
-        'rich-keypress-D',
-        'rich-beforeinput-D',
-        'rich-input-D',
-        'rich-keyup-D',
-    ];
-    let eventCounter = 0;
-
-    for (const targetId of ['plain', 'rich']) {
-        for (const eventType of ['beforeinput', 'input', 'keydown', 'keypress', 'keyup']) {
-            document.getElementById(targetId).addEventListener(eventType, t.step_func(event => {
-                if (event.key === 'Shift')
-                    return;
-                assert_equals(`${targetId}-${event.type}-${event.data || event.key}`, expectedResult[eventCounter]);
-                if (event instanceof InputEvent) {
-                    assert_equals(event.dataTransfer, null,
-                        'Plain text editing should have null |dataTransfer| field')
-                }
-                ++eventCounter;
-                if (eventCounter === expectedResult.length)
-                    t.done();
-            }));
+let inputEventsLog;
+
+function log(event) {
+    inputEventsLog.push({
+        type: event.type,
+        inputType: event.inputType,
+        data: event.data,
+        isComposing: event.isComposing,
+        cancelable: event.cancelable,
+        state: encodeURI(document.getElementById('rich').innerHTML)
+    });
+}
+
+function resetEditor(content = '') {
+    inputEventsLog = [];
+    document.getElementById('rich').innerHTML = content;
+}
+
+document.getElementById('rich').addEventListener('beforeinput', event => log(event));
+document.getElementById('rich').addEventListener('input', event => log(event));
+</script>
+<script>
+(function() {
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        const message = 'Hello';
+        await test_driver.send_keys(document.getElementById('rich'), message);
+        // 10 events (5 beforeinput + 5 input events)
+        assert_equals(inputEventsLog.length, 10);
+        for (let i = 0; i < inputEventsLog.length; i += 2) {
+            const beforeInputEvent = inputEventsLog[i];
+            const inputEvent = inputEventsLog[i + 1];
+            assert_equals(beforeInputEvent.type, 'beforeinput');
+            assert_equals(inputEvent.type, 'input');
+            assert_equals(beforeInputEvent.inputType, 'insertText');
+            assert_equals(inputEvent.inputType, 'insertText');
+            assert_equals(beforeInputEvent.data, inputEvent.data);
+            assert_equals(inputEvent.data, message[i / 2]);
```

Comment:
Maybe add a note here (or at the start of the look) that `i` is always being incremented by 2. Just so it's a bit more clear what is going on.

Reply:
I've added a comment. Not sure if it's where you thought it should be.